File: test/fortran/test_canonical.f90
program test_canon

use, intrinsic:: iso_fortran_env, only : stderr=>error_unit

use filesystem, only : path_t, get_cwd, same_file, canonical, is_dir, is_file, is_cygwin

implicit none


call test_canonical()
print '(a)', "OK: canonical full"

contains

subroutine test_canonical()

type(path_t) :: cur, par, file, p1, p2
character(*), parameter :: dummy = "nobody.txt"

integer :: L1, L2, L3

! -- current directory  -- old MacOS doesn't handle "." or ".." alone
cur = path_t(".")
cur = cur%canonical()
L1 = cur%length()
if (L1 < 1) then
  write(stderr,*) "ERROR: canonical '.' " // cur%path()
  error stop
endif

if (cur%path() /= get_cwd()) then
  write(stderr,*) "ERROR: canonical('.') " // cur%path() // " /= get_cwd: " // get_cwd()
  error stop
endif

print *, "OK: current dir = ", cur%path()
! -- home directory
p1 = path_t("~")
p1 = p1%canonical()
if (p1%path(1,1) == "~") error stop "%canonical ~ did not expanduser: " // p1%path()
if (canonical("~") == "~") error stop "canonical('~') should not be '~'"
print *, "OK: home dir = ", p1%path()

p2 = path_t(p1%parent())
L1 = p2%length()
if (L1 >= p1%length()) error stop "parent home " // p2%path()
print *, "OK: parent home = ", p2%path()


! -- relative dir
par = path_t("~/..")
par = par%canonical()

L2 = par%length()
if (L2 /= L1) then
  write(stderr,*) 'ERROR:canonical:relative: up dir not canonicalized: ~/.. => ' // par%path()
  error stop
endif
print *, 'OK: canon_dir = ', par%path()

! -- relative, non-existing file
if(is_cygwin()) then
  print '(a)', 'skip relative file not-exist as Cygwin does not support it'
else
file = path_t('~/../' // dummy)
file = file%canonical()
if(file%length() == 0) error stop "ERROR: relative file did not canonicalize: " // file%path()
L3 = file%length()

L1 = len(dummy)
if (L2 > 1) L1 = L1 + 1  !< when L2==1, $HOME is like /root instead of /home/user

if (L3 - L2 /= L1) then
  write(stderr,*) 'ERROR relative file was not canonicalized: ' // file%path(), L2, par%path(), L3, len(dummy)
  error stop
endif
endif !< is_cygwin

file = path_t('../' // dummy)
file = file%canonical()
if (file%path() /= "../" // dummy) then
  write(stderr,*) 'ERROR: relative file did not canonicalize: ' // file%path()
  error stop
end if

!> empty
if(canonical("") /= "") error stop "canonical('') " // canonical("") // " /= ''"

print *, 'OK: canon_file = ', file%path()

end subroutine test_canonical


end program

File: test/c/test_canonical.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef _MSC_VER
#include <crtdbg.h>
#endif

#include "ffilesystem.h"


int main(void){

#ifdef _MSC_VER
    _CrtSetReportMode(_CRT_ASSERT, _CRTDBG_MODE_FILE);
    _CrtSetReportFile(_CRT_ASSERT, _CRTDBG_FILE_STDERR);
    _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
    _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
    _CrtSetReportMode(_CRT_ERROR, _CRTDBG_MODE_FILE);
    _CrtSetReportFile(_CRT_ERROR, _CRTDBG_FILE_STDERR);
#endif

char out[FS_MAX_PATH];

// -- home directory
size_t L0 = fs_canonical("~", false, out, FS_MAX_PATH);

if(strcmp(out, "~") == 0){
  fprintf(stderr, "canonical(~) did not expanduser: %s\n", out);
  return EXIT_FAILURE;
}
printf("OK: home dir = %s\n", out);

size_t L1 = fs_parent(out, out, FS_MAX_PATH);
if (L1 >= L0)
  return EXIT_FAILURE;
printf("OK: parent home = %s\n", out);

// -- relative dir
const char* par = "~/..";

size_t L2 = fs_canonical(par, false, out, FS_MAX_PATH);

if (L2 != L1){
  fprintf(stderr, "ERROR:canonical:relative: up dir not canonicalized: ~/.. => %s\n", out);
  return EXIT_FAILURE;
}
printf("OK: canon_rel_up = %s\n", out);

// -- relative file
if(fs_is_cygwin())
  // Cygwin can't handle non-existing canonical paths
  return EXIT_SUCCESS;

const char* file = "~/../not-exist.txt";

size_t L = fs_canonical(file, false, out, FS_MAX_PATH);
if(L == 0) {
  fprintf(stderr, "ERROR: relative file did not resolve: %s\n", file);
  return EXIT_FAILURE;
}

size_t L4 = 13;
if(L2 > 1) L4++; // in case $HOME like /root instead of /home/user

if (L - L2 != L4){
  fprintf(stderr, "ERROR relative file was not canonicalized: %s %zu %zu\n", file, L0, L);
  return EXIT_FAILURE;
}

return EXIT_SUCCESS;
}
